controllers/reconciler.go

```diff
@@ -0,0 +1,336 @@
+// - Finalizers
+// - Status conditions
+// - Secrets (connection details)
+// - Events
+// - Requeue logic
+type Reconciler[T v1alpha1.AivenManagedObject] struct {
+	Controller
+	newController func(avnGen avngen.Client) AivenController[T]
+	newObj        func() T
+}
+
+// pollInterval controls how often we re-run reconciliation for resources that are in a steady state.
+// This enables continuous reconciliation without overwhelming the Aiven API.
+const pollInterval = 1 * time.Minute
+
+// Reconcile performs the full reconciliation loop for a managed resource.
+func (r *Reconciler[T]) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	obj := r.newObj()
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	ctx = logr.NewContext(ctx, setupLogger(r.Log, obj))
+
+	avnGen, err := r.newAivenClient(ctx, obj)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	controller := r.newController(avnGen)
+
+	if isMarkedForDeletion(obj) {
+		return r.finalize(ctx, controller, obj)
+	}
+
+	orig := obj.DeepCopyObject().(v1alpha1.AivenManagedObject)
+	defer func() {
+		err = errors.Join(err, r.updateStatus(ctx, orig, obj))
+	}()
+
+	if controllerutil.AddFinalizer(obj, instanceDeletionFinalizer) {
+		logr.FromContextOrDiscard(ctx).Info("added finalizer to instance")
+		r.Recorder.Event(obj, corev1.EventTypeNormal, eventAddedFinalizer, "instance finalizer added")
+	}
+
+	meta.SetStatusCondition(obj.Conditions(), getInitializedCondition("Preconditions", "Checking preconditions"))
+	obs, err := controller.Observe(ctx, obj)
+	if err != nil {
+		r.Recorder.Event(obj, corev1.EventTypeWarning, eventUnableToWaitForPreconditions, err.Error())
+		meta.SetStatusCondition(obj.Conditions(), getErrorCondition(errConditionPreconditions, err))
+		return ctrl.Result{}, fmt.Errorf("cannot observe the resource: %w", err)
+	}
+	if obs.PreconditionError != nil {
+		r.Recorder.Event(obj, corev1.EventTypeWarning, eventUnableToWaitForPreconditions, obs.PreconditionError.Error())
+		meta.SetStatusCondition(obj.Conditions(), getErrorCondition(errConditionPreconditions, obs.PreconditionError))
+		// TODO: Should we return an error here?
+		// TODO: Discuss if we need `obs.PreconditionError` at all, can we just return an error from Observe?
+		return ctrl.Result{RequeueAfter: requeueTimeout}, nil
+	}
+	if !obs.PreconditionsMet {
+		r.Recorder.Event(obj, corev1.EventTypeNormal, eventPreconditionsNotMet, "preconditions are not met, requeue")
+		logr.FromContextOrDiscard(ctx).Info("preconditions are not met, requeue")
+		return ctrl.Result{RequeueAfter: requeueTimeout}, nil
+	}
+	r.Recorder.Event(obj, corev1.EventTypeNormal, eventPreconditionsAreMet, "preconditions are met, proceeding to create or update")
+
+	if !obs.ResourceExists {
+		return r.createResource(ctx, controller, obj)
+	}
+
+	if !obs.ResourceUpToDate {
+		return r.updateResourceAndSecret(ctx, controller, obj)
+	}
```

Comment:
1. Let's check secret state separately.
2. When the secret is empty, do nothing, else update it.